### DISCUSSION TITLE
client error: forbidden for url

### DISCUSSION THREAD

**Discussion by woulschneider** [streamlink/streamlink#3912] [Q&A]
hi all, I get the following error when trying to download a m3u8 file from a site: 
(403 Client Error: Forbidden for url: ... )
 I am using the command :
 streamlink URL best
 can anyone help me ?

**Comment by mkbloke** [comment]
You will have to provide more information than that.  Include all the debug output provided when using the `-l debug` option.

**Reply by woulschneider** [reply]
streamlink -l debug https://cbbw-vh.akamaihd.net/i/5f7df07326ad6058b5169476/5f7df07326ad6058b5169476akamai.smil/master.m3u8?hdnea=exp=1628719977~acl=/*~hmac=609c163f6ea6169db19dc3de779b54ad2dbe1ac9de0c31ce438d165c2bd37c56
[cli][debug] OS:         Linux-5.11.0-25-generic-x86_64-with-glibc2.33
[cli][debug] Python:     3.9.5
[cli][debug] Streamlink: 2.2.0
[cli][debug] Requests(2.25.1), Socks(1.7.1), Websocket(0.57.0)
[cli][debug] Arguments:
[cli][debug]  url=https://cbbw-vh.akamaihd.net/i/5f7df07326ad6058b5169476/5f7df07326ad6058b5169476akamai.smil/master.m3u8?hdnea=exp=1628719977~acl=/*~hmac=609c163f6ea6169db19dc3de779b54ad2dbe1ac9de0c31ce438d165c2bd37c56
[cli][debug]  --loglevel=debug
[cli][info] Found matching plugin hls for URL https://cbbw-vh.akamaihd.net/i/5f7df07326ad6058b5169476/5f7df07326ad6058b5169476akamai.smil/master.m3u8?hdnea=exp=1628719977~acl=/*~hmac=609c163f6ea6169db19dc3de779b54ad2dbe1ac9de0c31ce438d165c2bd37c56
[plugins.hls][debug] URL=https://cbbw-vh.akamaihd.net/i/5f7df07326ad6058b5169476/5f7df07326ad6058b5169476akamai.smil/master.m3u8?hdnea=exp=1628719977~acl=/*~hmac=609c163f6ea6169db19dc3de779b54ad2dbe1ac9de0c31ce438d165c2bd37c56; params={}
[utils.l10n][debug] Language code: pt_BR
error: Unable to open URL: https://cbbw-vh.akamaihd.net/i/5f7df07326ad6058b5169476/5f7df07326ad6058b5169476akamai.smil/master.m3u8?hdnea=exp=1628719977~acl=/*~hmac=609c163f6ea6169db19dc3de779b54ad2dbe1ac9de0c31ce438d165c2bd37c56 (403 Client Error: Forbidden for url: https://cbbw-vh.akamaihd.net/i/5f7df07326ad6058b5169476/5f7df07326ad6058b5169476akamai.smil/master.m3u8?hdnea=exp=1628719977~acl=/*~hmac=609c163f6ea6169db19dc3de779b54ad2dbe1ac9de0c31ce438d165c2bd37c56)

**Reply by mkbloke** [reply]
You're using a tokenised link, which may or may not be valid at time of use.  Other possibilities are that it may be necessary to set the correct referer and/or origin headers in the request.  Where are you getting that m3u8 link from?

**Reply by woulschneider** [reply]
Hm...how do I add referrers ?
I get the link using web developers tools

**Reply by mkbloke** [reply]
I was asking about the source website URL rather than the method.

See the `--http-header` option.  Eg: `--http-header 'Referer=https://www.site.com/'`

**Reply by woulschneider** [reply]
oh, I believe that would be 'https://areaaluno.medcel.com.br/'. at least that was documented as the referrer url

**Reply by mkbloke** [reply]
> oh, I believe that would be 'https://areaaluno.medcel.com.br/'. at least that was documented as the referrer url

That requires a login, so I can't test if a recently acquired m3u8 link would work with, for example:
```
streamlink --http-header 'Referer=https://areaaluno.medcel.com.br/' --http-header 'Origin=https://areaaluno.medcel.com.br' <TokenisedURL> best
```
That's all I can suggest you try.

**Reply by woulschneider** [reply]
it worked !!
is there anyway I create a file with the stream ?

**Reply by mkbloke** [reply]
Use the `-o` option:
```
streamlink --http-header 'Referer=https://areaaluno.medcel.com.br/' --http-header 'Origin=https://areaaluno.medcel.com.br' -o <OutputFilename> <TokenisedURL> best
```

**Reply by woulschneider** [reply]
thank you ! it worked ! thank you very much !
